Updated RELEASES.md for 1.39.0 (#64878)
rust-lang/rust: RELEASES.md

Version 1.39.0 (2019-11-07)
===========================

Language
--------

> - [You can now create `async` functions and blocks with `async fn` and

Reviewer: ```suggestion
- [You can now create `async` functions and blocks with `async fn`, `async move {}`, and
```

  `async {}` respectively, and you can now call `.await` on async expressions.][63209]
- [You can now use `cfg` attributes on function parameters.][64010] e.g.
  ```rust
  fn len(
      #[cfg(windows)] slice: &[u16],
      #[cfg(not(windows))] slice: &[u8],
  ) -> usize {
      slice.len()
  }
  ```

> - [You can now bind patterns by move and bind by reference in `if` guards.][63118]

Reviewer: ```suggestion
- [You can now take references to bind-by-move patterns in the `if` guards of `match` arms.][63118]
```

  e.g.
  ```rust
  fn main() {
      let array: Box<[u8; 4]> = Box::new([1, 2, 3, 4]);

      match array {
          nums
  //      ---- `nums` is bound by move.
              if nums.iter().sum::<u8>() == 10
  //                 ^------ `.iter()` implicitly takes a reference to `nums`.
          => {
              drop(nums);
  //          ----------- Legal as `nums` was bound by move and so we have ownership.
          }
          _ => unreachable!(),
      }
  }
  ```

Compiler
--------
- [Added tier 3\* support for the `i686-unknown-uefi` target.][64334]
- [Added tier 3 support for the `sparc64-unknown-openbsd` target.][63595]

> - [rustc will now trim code snippets in diagnostics to fit in your terminal.][63402]

Reviewer 2: This does not apply when used with cargo, due to the way the two interact (see https://github.com/rust-lang/cargo/pull/7315).  I'm not sure if it should be called out if it won't work for most users who use cargo to build.

Reviewer: It should be pointed out yes since the compiler is an independent program. However, it is a good idea to point out what you just did in the text.

- [You can now pass `--show-output` argument to test binaries to print the
  stdout of successful tests.][62600]

\* Refer to Rust's [platform support page][forge-platform-support] for more
information on Rust's tiered platform support.

Libraries
---------

> **relnotes**

Reviewer 3: ```suggestion
```

- [`Vec::new` and `String::new` are now `const` functions.][64028]
- [`LinkedList::new` is now a `const` function.][63684]

> - [`asinh(-0.0)` now correctly returns `-0.0`.][63698] Previously this returned

Reviewer: Move to compatibility notes.

  `0.0`.
- [`str::len`, `[T]::len` and `str::as_bytes` are now `const` functions.][63770]
- [The `abs`, `wrapping_abs`, and `overflowing_abs` numeric functions are
  now `const`.][63786]

Stabilized APIs
---------------
- [`Pin::into_inner`]
- [`Instant::checked_duration_since`]
- [`Instant::saturating_duration_since`]

Cargo
-----
- [You can now publish git dependencies if supplied with a `version`.][cargo/7237]
- [The `--all` flag has been renamed to `--workspace`.][cargo/7241] Using
  `--all` is now deprecated.

Misc
----
- [You can now pass `-Clinker` to rustdoc to control the linker used
  for compiling doctests.][63834]

Compatibility Notes
-------------------
- [Code that was previously accepted by the old borrow checker, but rejected by
  the NLL borrow checker is now a hard error in Rust 2018.][63565] This was

>   previously a warning, and will become a hard error for Rust 2015 edition in a

Reviewer: ```suggestion
  previously a warning, and will become a hard error for Rust 2015 edition in 1.40.0.
```

>   future release.

Reviewer: ```suggestion
```

[64028]: https://github.com/rust-lang/rust/pull/64028/
[63595]: https://github.com/rust-lang/rust/pull/63595/
[63684]: https://github.com/rust-lang/rust/pull/63684/
[63698]: https://github.com/rust-lang/rust/pull/63698/
[63770]: https://github.com/rust-lang/rust/pull/63770/
[63786]: https://github.com/rust-lang/rust/pull/63786/
[63827]: https://github.com/rust-lang/rust/pull/63827/
[63834]: https://github.com/rust-lang/rust/pull/63834/
[63402]: https://github.com/rust-lang/rust/pull/63402/
[63565]: https://github.com/rust-lang/rust/pull/63565/
[63209]: https://github.com/rust-lang/rust/pull/63209/
[63118]: https://github.com/rust-lang/rust/pull/63118/
[62600]: https://github.com/rust-lang/rust/pull/62600/
[cargo/7237]: https://github.com/rust-lang/cargo/pull/7237/
[cargo/7241]: https://github.com/rust-lang/cargo/pull/7241/
[`Pin::into_inner`]: https://doc.rust-lang.org/std/pin/struct.Pin.html#method.into_inner
[`Instant::checked_duration_since`]: https://doc.rust-lang.org/std/time/struct.Instant.html#method.checked_duration_since
[`Instant::saturating_duration_since`]: https://doc.rust-lang.org/std/time/struct.Instant.html#method.saturating_duration_since

> 

Reviewer 4: ```suggestion
- The `try!` macro will now issue a deprecation warning.][62672] It is recommended to use the `?`
```

Version 1.39.0 (2019-11-07)
===========================

Language
--------
- [You can now create `async` functions and blocks with `async fn`, `async move {}`, and
  `async {}` respectively, and you can now call `.await` on async expressions.][63209]

> - [You can now use `cfg` attributes on function parameters.][64010] e.g.

Reviewer 5: This also applies to lint attributes, which I think is worth mentioning as well here

Version 1.39.0 (2019-11-07)
===========================

Language
--------
- [You can now create `async` functions and blocks with `async fn`, `async move {}`, and
  `async {}` respectively, and you can now call `.await` on async expressions.][63209]

> - [You can now use `cfg` and lint (e.g. `warn`, `deny`, `allow`) attributes on

Reviewer: You can also use inert attributes on function parameters that proc macros can use.

Author: How would you explain that briefly without confusing it with allowing proc macro attributes?

Reviewer: > You can now use certain attributes on function, closure, and function pointer parameters.
> These attributes include `cfg`, `cfg_attr`, `allow`, `warn`, `deny`, `forbid` as well as inert helper
> attributes used by procedural macro attributes applied to items.

Reviewer: cc @petrochenkov ^--

- [`LinkedList::new` is now a `const` function.][63684]
- [`str::len`, `[T]::len` and `str::as_bytes` are now `const` functions.][63770]
- [The `abs`, `wrapping_abs`, and `overflowing_abs` numeric functions are
  now `const`.][63786]

Stabilized APIs
---------------
- [`Pin::into_inner`]
- [`Instant::checked_duration_since`]
- [`Instant::saturating_duration_since`]

Cargo
-----
- [You can now publish git dependencies if supplied with a `version`.][cargo/7237]
- [The `--all` flag has been renamed to `--workspace`.][cargo/7241] Using
  `--all` is now deprecated.

Misc
----
- [You can now pass `-Clinker` to rustdoc to control the linker used
  for compiling doctests.][63834]

Compatibility Notes
-------------------
- [Code that was previously accepted by the old borrow checker, but rejected by
  the NLL borrow checker is now a hard error in Rust 2018.][63565] This was
  previously a warning, and will become a hard error for Rust 2015 edition in 1.40.0.
- [With some performance improvements when running large amount of doctests,
  `rustdoc` now requires `rustc` to be installed and in the same directory to
  run tests.][63827]

> - [The `try!` macro will now issue a deprecation warning.][62672] It is

Reviewer 6: Link `[62672]` is not referenced.

  ```rust
  fn len(
      #[cfg(windows)] slice: &[u16],
      #[cfg(not(windows))] slice: &[u8],
  ) -> usize {
      slice.len()
  }
  ```
- [You can now take references to bind-by-move patterns in the `if` guards of `match` arms.][63118]
  e.g.
  ```rust
  fn main() {
      let array: Box<[u8; 4]> = Box::new([1, 2, 3, 4]);

      match array {
          nums
  //      ---- `nums` is bound by move.
              if nums.iter().sum::<u8>() == 10
  //                 ^------ `.iter()` implicitly takes a reference to `nums`.
          => {
              drop(nums);
  //          ----------- Legal as `nums` was bound by move and so we have ownership.
          }
          _ => unreachable!(),
      }
  }
  ```

Compiler
--------
- [Added tier 3\* support for the `i686-unknown-uefi` target.][64334]
- [Added tier 3 support for the `sparc64-unknown-openbsd` target.][63595]
- [rustc will now trim code snippets in diagnostics to fit in your terminal.][63402]
  **Note** Cargo currently doesn't use this feature. Refer to
  [cargo#7315][cargo/7315] to track this feature's progress.
- [You can now pass `--show-output` argument to test binaries to print the

>   stdout of successful tests.][62600]

Reviewer 7: Seems like this bullet belongs in "libraries" (or maybe a new libtest section?)

\* Refer to Rust's [platform support page][forge-platform-support] for more
information on Rust's tiered platform support.

Libraries
---------
- [`Vec::new` and `String::new` are now `const` functions.][64028]
- [`LinkedList::new` is now a `const` function.][63684]
- [`str::len`, `[T]::len` and `str::as_bytes` are now `const` functions.][63770]
- [The `abs`, `wrapping_abs`, and `overflowing_abs` numeric functions are
  now `const`.][63786]

Stabilized APIs
---------------
- [`Pin::into_inner`]
- [`Instant::checked_duration_since`]
- [`Instant::saturating_duration_since`]

Cargo
-----
- [You can now publish git dependencies if supplied with a `version`.][cargo/7237]
- [The `--all` flag has been renamed to `--workspace`.][cargo/7241] Using
  `--all` is now deprecated.

Misc
----
- [You can now pass `-Clinker` to rustdoc to control the linker used
  for compiling doctests.][63834]

Compatibility Notes
-------------------
- [Code that was previously accepted by the old borrow checker, but rejected by
  the NLL borrow checker is now a hard error in Rust 2018.][63565] This was

>   previously a warning, and will become a hard error for Rust 2015 edition in 1.40.0.

Reviewer 7: ```suggestion
  previously a warning in the Rust 2015 edition, and will become a hard error in 1.40.0 in the 2015 and 2018 editions. It was already a hard error in the 2018 edition.
```

Version 1.39.0 (2019-11-07)
===========================

Language
--------
- [You can now create `async` functions and blocks with `async fn`, `async move {}`, and
  `async {}` respectively, and you can now call `.await` on async expressions.][63209]
- [You can now use certain attributes on function, closure, and function pointer
  parameters.][64010] These attributes include `cfg`, `cfg_attr`, `allow`, `warn`,
  `deny`, `forbid` as well as inert helper attributes used by procedural macro
  attributes applied to items. e.g.
  ```rust
  fn len(
      #[cfg(windows)] slice: &[u16],
      #[cfg(not(windows))] slice: &[u8],
  ) -> usize {
      slice.len()
  }
  ```

> - [You can now take references to bind-by-move patterns in the `if` guards of `match` arms.][63118]

Reviewer 8: Should this say `shared references` as to not be ambiguous, since `&mut` refs aren't allowed in the guards?

  //                 ^------ `.iter()` implicitly takes a reference to `nums`.
          => {
              drop(nums);
  //          ----------- Legal as `nums` was bound by move and so we have ownership.
          }
          _ => unreachable!(),
      }
  }
  ```

Compiler
--------
- [Added tier 3\* support for the `i686-unknown-uefi` target.][64334]
- [Added tier 3 support for the `sparc64-unknown-openbsd` target.][63595]
- [rustc will now trim code snippets in diagnostics to fit in your terminal.][63402]
  **Note** Cargo currently doesn't use this feature. Refer to
  [cargo#7315][cargo/7315] to track this feature's progress.

\* Refer to Rust's [platform support page][forge-platform-support] for more
information on Rust's tiered platform support.

Libraries
---------
- [`Vec::new` and `String::new` are now `const` functions.][64028]
- [`LinkedList::new` is now a `const` function.][63684]
- [`str::len`, `[T]::len` and `str::as_bytes` are now `const` functions.][63770]
- [The `abs`, `wrapping_abs`, and `overflowing_abs` numeric functions are
  now `const`.][63786]
- [You can now pass `--show-output` argument to test binaries to print the

>   stdout of successful tests.][62600]

Reviewer 7: I'm told in https://github.com/rust-lang/book/pull/2065 that this is also going to show stderr, so we should update this. Not sure if we should be more explicit or just say "output."

```suggestion
  output of successful tests.][62600]
```

Stabilized APIs
---------------
- [`Pin::into_inner`]
- [`Instant::checked_duration_since`]
- [`Instant::saturating_duration_since`]

Cargo
-----
- [You can now publish git dependencies if supplied with a `version`.][cargo/7237]
- [The `--all` flag has been renamed to `--workspace`.][cargo/7241] Using
  `--all` is now deprecated.

Misc
----
- [You can now pass `-Clinker` to rustdoc to control the linker used
  for compiling doctests.][63834]

Compatibility Notes
-------------------
- [Code that was previously accepted by the old borrow checker, but rejected by
  the NLL borrow checker is now a hard error in Rust 2018.][63565] This was

>   previously a warning in the Rust 2015 edition, and will become a hard error in 1.40.0 in the 2015 and 2018 editions. It was already a hard error in the 2018 edition.

Reviewer 9: This doesn't seem clear to me. First it says old borrowck bugs were previously warnings on 2018 and are now errors in 2018, then that they were already errors in 2018.

          nums
  //      ---- `nums` is bound by move.
              if nums.iter().sum::<u8>() == 10
  //                 ^------ `.iter()` implicitly takes a reference to `nums`.
          => {
              drop(nums);
  //          ----------- Legal as `nums` was bound by move and so we have ownership.
          }
          _ => unreachable!(),
      }
  }
  ```

Compiler
--------
- [Added tier 3\* support for the `i686-unknown-uefi` target.][64334]
- [Added tier 3 support for the `sparc64-unknown-openbsd` target.][63595]
- [rustc will now trim code snippets in diagnostics to fit in your terminal.][63402]
  **Note** Cargo currently doesn't use this feature. Refer to
  [cargo#7315][cargo/7315] to track this feature's progress.

\* Refer to Rust's [platform support page][forge-platform-support] for more
information on Rust's tiered platform support.

Libraries
---------
- [`Vec::new` and `String::new` are now `const` functions.][64028]
- [`LinkedList::new` is now a `const` function.][63684]
- [`str::len`, `[T]::len` and `str::as_bytes` are now `const` functions.][63770]
- [The `abs`, `wrapping_abs`, and `overflowing_abs` numeric functions are
  now `const`.][63786]

> - [You can now pass `--show-output` argument to test binaries to print the

Reviewer: This should be moved to # Compiler

  output of successful tests.][62600]

Stabilized APIs
---------------
- [`Pin::into_inner`]
- [`Instant::checked_duration_since`]
- [`Instant::saturating_duration_since`]

Cargo
-----
- [You can now publish git dependencies if supplied with a `version`.][cargo/7237]
- [The `--all` flag has been renamed to `--workspace`.][cargo/7241] Using
  `--all` is now deprecated.

Misc
----
- [You can now pass `-Clinker` to rustdoc to control the linker used
  for compiling doctests.][63834]

Compatibility Notes
-------------------
- [Code that was previously accepted by the old borrow checker, but rejected by
  the NLL borrow checker is now a hard error in Rust 2018.][63565] This was
  previously a warning in the Rust 2015 edition, and will become a hard error

>   in the 1.40.0 release.

Reviewer 10: This wording still seems confusing. The link talks about Rust 2018, but the second sentence talks about Rust 2015 only with no link to anything?! So is it just Rust 2018, just Rust 2015 or both?

- [With some performance improvements when running large amount of doctests,
  `rustdoc` now requires `rustc` to be installed and in the same directory to

>   run tests.][63827]

Reviewer 11: I have a hard time parsing this. It should probably be broken up, like

> `rustdoc` now requires `rustc` to be installed and in the same directory to run tests. Performance was improved when running a large amount of doctests.
